sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerSchema.java

```diff
@@ -19,68 +19,105 @@
+  abstract ImmutableListMultimap<String, KeyPart> keyParts();
+  abstract ImmutableTable<String, String, Long> cellsMutatedPerColumn();
+  abstract ImmutableMap<String, Long> cellsMutatedPerRow();
 
   public static Builder builder() {
-    return new Builder();
+    return new AutoValue_SpannerSchema.Builder();
   }
 
   /**
    * Builder for {@link SpannerSchema}.
    */
-  static class Builder {
-    private final ArrayListMultimap<String, Column> columns = ArrayListMultimap.create();
-    private final ArrayListMultimap<String, KeyPart> keyParts = ArrayListMultimap.create();
-
+  @AutoValue.Builder
+  abstract static class Builder {
+    abstract ImmutableList.Builder<String> tablesBuilder();
+    abstract ImmutableListMultimap.Builder<String, Column> columnsBuilder();
+    abstract ImmutableListMultimap.Builder<String, KeyPart> keyPartsBuilder();
+    abstract ImmutableTable.Builder<String, String, Long> cellsMutatedPerColumnBuilder();
+    abstract ImmutableMap.Builder<String, Long> cellsMutatedPerRowBuilder();
+
+    abstract ImmutableListMultimap<String, Column> columns();
+    abstract ImmutableTable<String, String, Long> cellsMutatedPerColumn();
+
+    @VisibleForTesting
     public Builder addColumn(String table, String name, String type) {
-      addColumn(table, Column.create(name.toLowerCase(), type));
-      return this;
+      return addColumn(table, name, type, 1L);
     }
 
-    private Builder addColumn(String table, Column column) {
-      columns.put(table.toLowerCase(), column);
+    public Builder addColumn(String table, String name, String type, long cellsMutated) {
+      String tableLower = table.toLowerCase();
+      String nameLower = name.toLowerCase();
+
+      columnsBuilder().put(tableLower, Column.create(nameLower, type));
+      cellsMutatedPerColumnBuilder().put(tableLower, nameLower, cellsMutated);
       return this;
     }
 
     public Builder addKeyPart(String table, String column, boolean desc) {
-      keyParts.put(table.toLowerCase(), KeyPart.create(column.toLowerCase(), desc));
+      keyPartsBuilder().put(table.toLowerCase(), KeyPart.create(column.toLowerCase(), desc));
       return this;
     }
 
-    public SpannerSchema build() {
-      return new SpannerSchema(columns, keyParts);
-    }
-  }
+    abstract SpannerSchema autoBuild();
+
+    public final SpannerSchema build() {
+      // precompute the number of cells that are mutated for operations affecting
+      // an entire row such as a single key delete.
+      cellsMutatedPerRowBuilder().putAll(Maps.transformValues(
+          cellsMutatedPerColumn().rowMap(),
+          entry -> entry.values().stream().mapToLong(Long::longValue).sum()));
 
-  private SpannerSchema(ArrayListMultimap<String, Column> columns,
-      ArrayListMultimap<String, KeyPart> keyParts) {
-    this.columns = columns;
-    this.keyParts = keyParts;
-    tables = new ArrayList<>(columns.keySet());
+      tablesBuilder().addAll(columns().keySet());
```

Comment:
This makes the method non-idempotent. Is there a way to refactor this to be simpler, and not have the builder method mutate the value?

Reply:
We can just use a setter here. Updated.